<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Allelic frequency — ds.alleleFrequency • dsOmicsClient</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Allelic frequency — ds.alleleFrequency"><meta property="og:description" content="Calculates the frequency of the A allele on a server side GenotypeData object."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">dsOmicsClient</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.8</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/geno_exploration.html">OmicSHIELD bookdown</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Allelic frequency</h1>
    
    <div class="hidden name"><code>ds.alleleFrequency.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Calculates the frequency of the A allele on a server side GenotypeData object.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">ds.alleleFrequency</span><span class="op">(</span>
  <span class="va">genoData</span>,
  type <span class="op">=</span> <span class="st">"combined"</span>,
  method <span class="op">=</span> <span class="st">"fast"</span>,
  snpBlock <span class="op">=</span> <span class="fl">5000L</span>,
  datasources <span class="op">=</span> <span class="cn">NULL</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>genoData</dt>
<dd><p><code>character</code> Name of the <code>GenotypeData</code> object on the server</p></dd>
<dt>type</dt>
<dd><p><code>character</code> (default <code>"combined"</code>) Type of analysis, if (<code>"split"</code>), the
frequencies will be calculated for each study server. If  (<code>"combined"</code>) 
a pooled methodology will be performed.</p></dd>
<dt>method</dt>
<dd><p><code>character</code> (default <code>"fast"</code>) If <code>"fast"</code> an optimized fast method will 
be used; if <code>"slow"</code> the function <code><a href="https://rdrr.io/pkg/GWASTools/man/alleleFrequency.html" class="external-link">GWASTools::alleleFrequency</a></code> will be used, this is notably 
slower.</p></dd>
<dt>snpBlock</dt>
<dd><p><code>integer</code> (default <code>5000L</code>) number of SNPs to read at each iteration, 
tune this parameter to improve performance. This argument is only used when <code>method = "fast"</code></p></dd>
<dt>datasources</dt>
<dd><p>a list of <code>DSConnection-class</code> objects obtained after login.</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A <code>data frame</code> with a row for each SNP. Columns "M" for males, "F" for females, and "all" 
for all scans give frequencies of the A allele. Sample size for males, females, and all is returned as 
"n.M", "n.F", and "n", respectively. "MAF" is the minor allele frequency over all scans.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>(From alleleFrequency documentation): Counts male heterozygotes on the X and Y chromosomes
 as missing values, and any genotype for females on the Y chromosome as missing values. A "sex" 
 variable must be present in the scan annotation slot of genoData.
 Samples with missing sex are included in the allele counts for "all" and "MAF" for autosomes, 
 but not for sex chromosomes.</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Yannick Marcon, Xavier Esriba-Montagut, Juan R. Gonzalez.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

