<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ dsOmicsClient</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">dsOmicsClient</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.8</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/geno_exploration.html">OmicSHIELD bookdown</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="checkForExpressionSet.html">checkForExpressionSet()</a></code> </p>
        </td>
        <td><p>Check if object exists on the study server and if it's an ExpressionSet</p></td>
      </tr><tr><td>
          <p><code><a href="dot-eafPlot.html">.eafPlot()</a></code> </p>
        </td>
        <td><p>Auxiliary function for <code>eafPlot</code></p></td>
      </tr><tr><td>
          <p><code><a href="dot-pzPlotgg.html">.pzPlotgg()</a></code> </p>
        </td>
        <td><p>Auxiliary function for <code>pzPlot</code></p></td>
      </tr><tr><td>
          <p><code><a href="ds.addPhenoData2eSet.html">ds.addPhenoData2eSet()</a></code> </p>
        </td>
        <td><p>Add Phenotype data to ExpressionSet</p></td>
      </tr><tr><td>
          <p><code><a href="ds.alleleFrequency.html">ds.alleleFrequency()</a></code> </p>
        </td>
        <td><p>Allelic frequency</p></td>
      </tr><tr><td>
          <p><code><a href="ds.createRSE.html">ds.createRSE()</a></code> </p>
        </td>
        <td><p>Create a RangedSummarizedExperiment from a RNAseq count table and 
a phenotypes table</p></td>
      </tr><tr><td>
          <p><code><a href="ds.DESeq2.html">ds.DESeq2()</a></code> </p>
        </td>
        <td><p>Server-side Differential Gene Expression analysis using DESeq2</p></td>
      </tr><tr><td>
          <p><code><a href="ds.edgeR.html">ds.edgeR()</a></code> </p>
        </td>
        <td><p>Server-side Differential Expression Analysis using edgeR</p></td>
      </tr><tr><td>
          <p><code><a href="ds.exactHWE.html">ds.exactHWE()</a></code> </p>
        </td>
        <td><p>Hardy-Weinberg Equilibrium testing</p></td>
      </tr><tr><td>
          <p><code><a href="ds.fastGWAS.html">ds.fastGWAS()</a></code> </p>
        </td>
        <td><p>Fast genome-wide association analysis (GWAS)</p></td>
      </tr><tr><td>
          <p><code><a href="ds.featureData.html">ds.featureData()</a></code> </p>
        </td>
        <td><p>Retrieve information on features from eSets and RSEs.</p></td>
      </tr><tr><td>
          <p><code><a href="ds.featureNames.html">ds.featureNames()</a></code> </p>
        </td>
        <td><p>Retrieve feature names from eSets and RSEs.</p></td>
      </tr><tr><td>
          <p><code><a href="ds.fvarLabels.html">ds.fvarLabels()</a></code> </p>
        </td>
        <td><p>Retrieve phenotypic data (e.g. covariates) from eSets and RSEs.</p></td>
      </tr><tr><td>
          <p><code><a href="ds.genoDimensions.html">ds.genoDimensions()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="ds.GenotypeData.html">ds.GenotypeData()</a></code> </p>
        </td>
        <td><p>Creates a GenotypeData object at each server</p></td>
      </tr><tr><td>
          <p><code><a href="ds.getChromosomeNames.html">ds.getChromosomeNames()</a></code> </p>
        </td>
        <td><p>Get names of chromosomes</p></td>
      </tr><tr><td>
          <p><code><a href="ds.getSNPs.html">ds.getSNPs()</a></code> </p>
        </td>
        <td><p>Retrieve SNPs from</p></td>
      </tr><tr><td>
          <p><code><a href="ds.getSNPSbyGen.html">ds.getSNPSbyGen()</a></code> </p>
        </td>
        <td><p>Subset GDS with gene(s)</p></td>
      </tr><tr><td>
          <p><code><a href="ds.glmSNP.html">ds.glmSNP()</a></code> </p>
        </td>
        <td><p>Logistic regression analysis of pooled data for each SNP site in study</p></td>
      </tr><tr><td>
          <p><code><a href="ds.index.gdsn.html">ds.index.gdsn()</a></code> </p>
        </td>
        <td><p>Retrieve phenotypic data (e.g. covariates) from eSets.</p></td>
      </tr><tr><td>
          <p><code><a href="ds.limma.html">ds.limma()</a></code> </p>
        </td>
        <td><p>Server-side Differential Gene Expression analysis using limma</p></td>
      </tr><tr><td>
          <p><code><a href="ds.lmFeature.html">ds.lmFeature()</a></code> </p>
        </td>
        <td><p>Linear regression analysis of pooled data for each CpG site in study</p></td>
      </tr><tr><td>
          <p><code><a href="ds.make_valid_names.html">ds.make_valid_names()</a></code> </p>
        </td>
        <td><p>Rename column names and character variables</p></td>
      </tr><tr><td>
          <p><code><a href="ds.metaGWAS.html">ds.metaGWAS()</a></code> </p>
        </td>
        <td><p>Genome-wide association analysis (GWAS)</p></td>
      </tr><tr><td>
          <p><code><a href="ds.nFeatures.html">ds.nFeatures()</a></code> </p>
        </td>
        <td><p>Retrieve the number of features for eSets and RSEs.</p></td>
      </tr><tr><td>
          <p><code><a href="ds.nSamples.html">ds.nSamples()</a></code> </p>
        </td>
        <td><p>Retrieve the number of samples for eSets and RSEs.</p></td>
      </tr><tr><td>
          <p><code><a href="ds.PCA.html">ds.PCA()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="ds.PCASNPS.html">ds.PCASNPS()</a></code> </p>
        </td>
        <td><p>Principal Component Analysis (PCA) on SNP genotype data</p></td>
      </tr><tr><td>
          <p><code><a href="ds.PLINK.html">ds.PLINK()</a></code> </p>
        </td>
        <td><p>Genome-wide association analysis (GWAS) using PLINK</p></td>
      </tr><tr><td>
          <p><code><a href="ds.PRS.html">ds.PRS()</a></code> </p>
        </td>
        <td><p>Get Ploygenic Risk Score</p></td>
      </tr><tr><td>
          <p><code><a href="ds.removeOutliers.html">ds.removeOutliers()</a></code> </p>
        </td>
        <td><p>Filter potential CpG outliers</p></td>
      </tr><tr><td>
          <p><code><a href="ds.RNAseqPreproc.html">ds.RNAseqPreproc()</a></code> </p>
        </td>
        <td><p>Filter Genes By Expression Level</p></td>
      </tr><tr><td>
          <p><code><a href="ds.snptest.html">ds.snptest()</a></code> </p>
        </td>
        <td><p>Interface to run SNPtest commands on a ssh connection (with SNPtest 2.5.2 installed)</p></td>
      </tr><tr><td>
          <p><code><a href="ds.subsetExpressionSet.html">ds.subsetExpressionSet()</a></code> </p>
        </td>
        <td><p>Subset ExpressionSet</p></td>
      </tr><tr><td>
          <p><code><a href="ds.table_gds.html">ds.table_gds()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="ds.varLabels.html">ds.varLabels()</a></code> </p>
        </td>
        <td><p>Retrieve phenotypic data (e.g. covariates) from eSets.</p></td>
      </tr><tr><td>
          <p><code><a href="eafPlot.html">eafPlot()</a></code> </p>
        </td>
        <td><p>EAF plot</p></td>
      </tr><tr><td>
          <p><code><a href="get_EGA_url.html">get_EGA_url()</a></code> </p>
        </td>
        <td><p>Get EGA URL</p></td>
      </tr><tr><td>
          <p><code><a href="get_ga4gh_url.html">get_ga4gh_url()</a></code> </p>
        </td>
        <td><p>Get GA4GH URL</p></td>
      </tr><tr><td>
          <p><code><a href="glmSNP.html">glmSNP()</a></code> </p>
        </td>
        <td><p>Fits a GLM to assess association between the outcome (binary variable) and a given SNP</p></td>
      </tr><tr><td>
          <p><code><a href="lambdaNPlot.html">lambdaNPlot()</a></code> </p>
        </td>
        <td><p>Lambda-N plot</p></td>
      </tr><tr><td>
          <p><code><a href="lmFeature.html">lmFeature()</a></code> </p>
        </td>
        <td><p>Fits a LM to assess association between the features (outcome) and grouping variable (e.g case/control, condition, ...)</p></td>
      </tr><tr><td>
          <p><code><a href="LocusZoom.html">LocusZoom()</a></code> </p>
        </td>
        <td><p>LocusZoom</p></td>
      </tr><tr><td>
          <p><code><a href="manhattan.html">manhattan()</a></code> </p>
        </td>
        <td><p>Mannhattan Plot</p></td>
      </tr><tr><td>
          <p><code><a href="metaBetaValues.html">metaBetaValues()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="metaPvalues.html">metaPvalues()</a></code> </p>
        </td>
        <td><p>Meta-analysis of p-values</p></td>
      </tr><tr><td>
          <p><code><a href="plot.dsLmFeature.html">plot(<i>&lt;dsLmFeature&gt;</i>)</a></code> </p>
        </td>
        <td><p>Generate a QQ-plot of p-values output from any omic data analysis</p></td>
      </tr><tr><td>
          <p><code><a href="plotPCA.html">plotPCA()</a></code> </p>
        </td>
        <td><p>Plot the results of a <code><a href="../reference/ds.PCA.html">ds.PCA()</a></code></p></td>
      </tr><tr><td>
          <p><code><a href="plotPCASNPS.html">plotPCASNPS()</a></code> </p>
        </td>
        <td><p>Plot the results of a <code><a href="../reference/ds.PCASNPS.html">ds.PCASNPS()</a></code></p></td>
      </tr><tr><td>
          <p><code><a href="pzPlot.html">pzPlot()</a></code> </p>
        </td>
        <td><p>P-Z plot</p></td>
      </tr><tr><td>
          <p><code><a href="qqplot.html">qqplot()</a></code> </p>
        </td>
        <td><p>Create a quantile-quantile plot with ggplot2.</p></td>
      </tr><tr><td>
          <p><code><a href="seNPlot.html">seNPlot()</a></code> </p>
        </td>
        <td><p>SE-N Plot</p></td>
      </tr><tr><td>
          <p><code><a href="standardizeGenoData.html">standardizeGenoData()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="tolist.html">tolist()</a></code> </p>
        </td>
        <td><p>PLINK arguments in the DataSHIELD required format</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Yannick Marcon, Xavier Esriba-Montagut, Juan R. Gonzalez.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

