<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>LocusZoom — LocusZoom • dsOmicsClient</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="LocusZoom — LocusZoom"><meta property="og:description" content="LocusZoom"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">dsOmicsClient</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.8</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/geno_exploration.html">OmicSHIELD bookdown</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>LocusZoom</h1>
    
    <div class="hidden name"><code>LocusZoom.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>LocusZoom</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">LocusZoom</span><span class="op">(</span>
  <span class="va">gwas_results</span>,
  range <span class="op">=</span> <span class="fl">5e+05</span>,
  position_zoom <span class="op">=</span> <span class="cn">NULL</span>,
  use_biomaRt <span class="op">=</span> <span class="cn">TRUE</span>,
  TxDb_version <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"hg18"</span>, <span class="st">"hg19"</span><span class="op">)</span>,
  GRCh <span class="op">=</span> <span class="cn">NULL</span>,
  pvalue <span class="op">=</span> <span class="st">"p.value"</span>,
  chromosome <span class="op">=</span> <span class="st">"chr"</span>,
  position <span class="op">=</span> <span class="st">"pos"</span>,
  rs <span class="op">=</span> <span class="st">"rs"</span>,
  range_filter <span class="op">=</span> <span class="cn">NULL</span>,
  draw_genes <span class="op">=</span> <span class="cn">TRUE</span>,
  pval_threshold <span class="op">=</span> <span class="fl">5</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>gwas_results</dt>
<dd><p><code>data.frame</code> With atleast three columns: Chromosome, position and p-value</p></dd>
<dt>range</dt>
<dd><p><code>numeric</code> (default <code>5e+05</code>) Range of the plot in genome positions.</p></dd>
<dt>position_zoom</dt>
<dd><p><code>numeric</code> (default <code>NULL</code>) Center position of the plot if supplied. If <code>NULL</code>, 
the center position is the lowerst p-value of the <code>gwas_results</code> table.</p></dd>
<dt>use_biomaRt</dt>
<dd><p><code>bool</code> (default <code>TRUE</code>) Set to <code>TRUE</code> to use biomaRt to retrieve the annotation of the 
genes. If set to <code>FALSE</code> TxDb.Hsapiens.UCSC.hgXX.knownGene will be used. Use the argument <code>TxDb_version</code> to 
select between hg19 or hg18. Note that when using TxDb the correspondent packages needs to be installed, providing 
a internet-free annotation option.</p></dd>
<dt>TxDb_version</dt>
<dd><p><code>character</code> (default <code>hg18</code>) TxDb version to use. This argument is only used when 
<code>use_biomaRt</code> is set to <code>FALSE</code>.</p></dd>
<dt>GRCh</dt>
<dd><p><code>numeric</code> (default <code>NULL</code>) GRCh version to connect to if not the current GRCh38, currently this can 
only be 37, if is different from 37 the GRCh38 version will be used. (Parameter passed to biomaRt::useEnsembl). 
This argument is only used when <code>use_biomaRt</code> is set to TRUE.</p></dd>
<dt>pvalue</dt>
<dd><p><code>character</code> (default <code>"p.value"</code>) Column name of <code>gwas_results</code> that contains the p-values.</p></dd>
<dt>chromosome</dt>
<dd><p><code>character</code> (default <code>"chr"</code>) Column name of <code>gwas_results</code> that contains 
the chromosome names.</p></dd>
<dt>position</dt>
<dd><p><code>character</code> (default <code>"pos"</code>) Column name of <code>gwas_results</code> that contains the positions.</p></dd>
<dt>rs</dt>
<dd><p><code>character</code> (default <code>"rs"</code>) Column name of <code>gwas_results</code> that contains 
the rs ID of the SNPs. This column will only be used if <code>pval_threshold</code> is different than <code>NULL</code>, therefore 
there is no need of supplying this argument if there is no interest to plot the labels of the rs IDs.</p></dd>
<dt>range_filter</dt>
<dd><p><code>numeric</code> (default <code>NULL</code>). If not <code>NULL</code>, only the annotations with a range larger 
than <code>range_filter</code> will be displayed.</p></dd>
<dt>draw_genes</dt>
<dd><p><code>bool</code> (default <code>TRUE</code>). Select if the genes of the region plot are to be plotted below the p-values 
plot.</p></dd>
<dt>pval_threshold</dt>
<dd><p><code>numeric</code> (default <code>5</code>) Threshold of -log_10p-value to draw labels with the SNP IDs. 
If <code>NULL</code> no labels will be drawn.</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Yannick Marcon, Xavier Esriba-Montagut, Juan R. Gonzalez.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

